light/client.go

```diff
@@ -879,20 +905,64 @@ func (c *Client) updateTrustedHeaderAndVals(h *types.SignedHeader, vals *types.V
 	return nil
 }
 
-// fetch header and validators for the given height (0 - latest) from primary
-// provider.
-func (c *Client) fetchHeaderAndValsAtHeight(height int64) (*types.SignedHeader, *types.ValidatorSet, error) {
+// 0 - latest header
+// Note it swaps the primary with a witness if primary is not responding after
+// MaxRetryAttempts.
+func (c *Client) signedHeaderAndValSetFromPrimary(height int64) (*types.SignedHeader, *types.ValidatorSet, error) {
 	h, err := c.signedHeaderFromPrimary(height)
```

Comment:
As an idea, it might be a lot cleaner if instead of having `c.signedHeaderFromPrimary` we had `c.signedHeaderFrom` and then we wouldn't need two functions for signedHeaderAndValSetFromPrimary and signedHeaderAndValSetFromWitness but just have `signedHeaderAndValSetFrom` pass the provider on to the new  created `c.signedHeaderFrom` - I don't know if you thought about this before but it may reduce the amount of code we need and make it cleaner - just an idea though

Reply:
I thought about it before. The difference between the two is signedHeaderFromPrimary does multiple retries and replaces primary on failures. signedHeaderFromWitness does not retry nor we need to replace it. I think this justifies two distinct functions

Reply:
yeah I see, don't we do exponential backoff retries when we are requesting headers from witnesses (we just don't remove them)